app/src/lib/parse-carriage-return.ts

```diff
@@ -0,0 +1,50 @@
+/**
+ * Parses carriage returns the same way a terminal would, i.e by
+ * moving the cursor and (potentially) overwriting text.
+ *
+ * Git (and many other CLI tools) use this trick to present the
+ * user with nice looking progress. When writing something like...
+ *
+ * 'Downloading: 1%  \r'
+ * 'Downloading: 2%  \r'
+ *
+ * ...to the terminal he user is gonna perceieve it as if the 1 just
+ * magically changes to a two.
+ *
+ * The carriage return character for all of you kids out there
+ * that haven't yet played with a manual typewriter refers to the
+ * "carriage" which held the character arms, see
+ *
+ *  https://en.wikipedia.org/wiki/Carriage_return#Typewriters
+ */
+export function parseCarriageReturn(text: string) {
+  // Happy path, there are no carriage returns in
+  // the text, making this method a noop.
+  if (text.indexOf('\r') < 0) {
+    return text
+  }
+
+  const lines = new Array<string>('')
+  const crLfOrEnd = /(.*?)([\r\n$])/gm
```

Comment:
Do we need the question mark on the RegExp?

```suggestion
  const crLfOrEnd = /(.*)([\r\n$])/gm
```

Reply:
Technically no since `.` doesn't match newline characters in javascript, but in other languages we would have had to in order to make the expression lazy to ensure it matches as few times as possible. If we made it greedy the first group could end up eating up one or `\r` or `\n` if they are adjacent to other newline characters.

I think being explicit about the need for the first expression to be lazy here is a good thing though so I'd like to keep it that way.

Reply:
Thanks for the context!! Super useful!

> but in other languages we would have had to in order to make the expression lazy

I thought that in all RegExp implementations newlines were not matched by `.` 😅  (too much time doing JS 🙃).

I agree about keeping it but I would add a comment about why it's there (otherwise I can see my future self removing it LOL).

BTW I just noticed the `$` sign in the last group, I think that it should be moved to the end of the expression since afaik right now it will match the dollar character.

```suggestion
  const crLfOrEnd = /(.*?)([\r\n])$/gm
```

Reply:
> I thought that in all RegExp implementations newlines were not matched by . 😅 (too much time doing JS 🙃).

And I learned that my knowledge was outdated here as well, turns out [`.` matching newlines](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/dotAll) has been supported in most browser for [a good while now](https://caniuse.com/#search=dotall).

> BTW I just noticed the $ sign in the last group, I think that it should be moved to the end of the expression since afaik right now it will match the dollar character.

Wow, I don't know how I missed that, and even more depressing is that I hadn't written a test to catch it. I've added both a test and changed the logic now.

> otherwise I can see my future self removing it LOL

Soooo... as I was researching the dotAll stuff I happened to see that JavaScript `.` matches on unicode line and paragraph breaks as well 😱. I feel like I knew this at the back of my head but have never really had to care about it. Now we have to though because hypothetically the remote could send either of those characters so I've had added a test specifically for that and updated the RegExp to use `[\s\S]` as a more explicit substitute for `.` i.e. we want it to match _EVERYTHING_ (but lazily)

Reply:
Wow! Really good catch and thanks for adding the super detailed comment!

![image](https://user-images.githubusercontent.com/408035/84142205-ab87cd00-aa54-11ea-93ae-22afdaf0afcb.png)